### PR TITLE
Support cyclic union types

### LINKED ISSUE #22429: Remove error from anydata
https://github.com/ballerina-platform/ballerina-spec/issues/463

### LINKED ISSUE #23691: Update param and return types for clone, cloneReadOnly
**Description:**
Please refer https://github.com/ballerina-platform/ballerina-spec/issues/531#issuecomment-635739590

The clone, and cloneReadOnly should be changed to 
```ballerina
type Cloneable readonly|xml|Cloneable[]|map<Cloneable>|table<map<Cloneable>>;

@typeParam
type CloneableType Cloneable;

public function clone(CloneableType v) returns CloneableType = external;

public function cloneReadOnly(CloneableType v) returns CloneableType & readonly = external;
```
This depends on https://github.com/ballerina-platform/ballerina-lang/issues/23690

### LINKED ISSUE #24337: Replace error detail type from (anydata|readonly) to Cloneable
**Description:**
Currently the detail field in error is of type `(anydata|error)`. This should be replaced with `Cloneable` type.

Please refer issue: https://github.com/ballerina-platform/ballerina-spec/issues/531

https://github.com/ballerina-platform/ballerina-spec/commit/4aeca2a0f9ccaa29f250f6900284cdd709a99a21#diff-5d269469223332ba2a1db22eddc06d7fL19

### LINKED ISSUE #24338: Replace static type of send action from (anydata|readonly) to Cloneable
**Description:**
Currently the static type of send action has the type of `(anydata|readonly)`. This needs to replaced with `Cloneable`.

Related spec issue: https://github.com/ballerina-platform/ballerina-spec/issues/531

https://github.com/ballerina-platform/ballerina-spec/commit/4aeca2a0f9ccaa29f250f6900284cdd709a99a21#diff-a9b79438f2c8a8890efc4088543d5a35L6040

### LINKED ISSUE #24339: Replace type of annotation tag from (anydata|readonly) to Cloanable
**Description:**
Currently the annotation tags are of type `anydata|readonly`, this needs to be replaced with `Cloanable`

Related issue: https://github.com/ballerina-platform/ballerina-spec/issues/531

https://github.com/ballerina-platform/ballerina-spec/commit/4aeca2a0f9ccaa29f250f6900284cdd709a99a21#diff-a9b79438f2c8a8890efc4088543d5a35L7625-L7627

### PR DESCRIPTION
**Pull request by KRVPerera**
## Purpose
part of #23691 - Improve support for clone, cloneReadOnly
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/23690 - Allow cyclic reference in unions typedef
Fixes #22429 - Remove error from anydata
Fixes #24337 - Replace error detail type from (anydata|readonly) to Cloneable
Fixes #24339 - Replace type of annotation tag from (anydata|readonly) to Cloanable 
Fixes #24338 - Replace static type of send action from (anydata|readonly) to Cloneable

## Remarks
Done a hack to finish the implementation.
https://github.com/ballerina-platform/ballerina-lang/issues/27786 TypeParam flag is propagated to original type 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by KRVPerera** [comment]
Create an issue for the PR for 
```Ballerina
@typeParam
type CloneableType Cloneable|never;		// hack
//type CloneableType Cloneable;			// expected
```
https://github.com/ballerina-platform/ballerina-lang/issues/27786

File: bvm/ballerina-core/src/main/java/org/ballerinalang/core/model/types/BUnionType.java [hunk 1]
```diff
@@ -91,8 +92,27 @@ public int getTag() {
...
     @Override
     public String toString() {
-        List<String> list = memberTypes.stream().map(BType::toString).collect(Collectors.toList());
-        return String.join("|", list);
+        String cachedToString;
+        if (resolving) {
+            if (this.typeName != null) {
+                return this.typeName;
+            } else {
+                return "...";
```
Comment (MaryamZi) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/core/model/types/BUnionType.java:100]: Can `this.typeName` be null when `resolving` is true? Wouldn't `resolving` be true only for a cyclic union, which in turn has to have a name?

File: bvm/ballerina-core/src/main/java/org/ballerinalang/core/model/types/BUnionType.java [hunk 2]
```diff
@@ -91,8 +92,27 @@ public int getTag() {
...
+            }
+        }
+        resolving = true;
+        StringBuilder sb = new StringBuilder();
+        int size = memberTypes != null ? memberTypes.size() : 0;
```
Comment (MaryamZi) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/core/model/types/BUnionType.java:105]: Can `toString` get called when `memberTypes` is null?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/PredefinedTypes.java [hunk 3]
```diff
@@ -163,17 +159,82 @@
...
+        members.add(TYPE_DECIMAL);
+        members.add(TYPE_STRING);
+        members.add(TYPE_XML);
+        BAnydataType anydataType = new BAnydataType(new BUnionType(TypeConstants.ANYDATA_TNAME, EMPTY_MODULE,
+                members), TypeConstants.ANYDATA_TNAME, false);
+        anydataType.isCyclic = true;
+        MapType internalAnydataMap = new BMapType(TypeConstants.MAP_TNAME, anydataType, EMPTY_MODULE);
+        ArrayType internalAnydataArray = new BArrayType(anydataType);
+        members.add(internalAnydataArray);
+        members.add(internalAnydataMap);
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/PredefinedTypes.java:195]: What about `table<map<anydata>>`?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 4]
```diff
@@ -2862,44 +2875,49 @@ private static boolean hasFillerValue(Type type, List<Type> unanalyzedTypes) {
...
-        Type firstMember;
-        for (firstMember = iterator.next(); iterator.hasNext(); ) {
-            if (!isSameType(firstMember, iterator.next())) {
-                return false;
-            }
-        }
-        // Control reaching this point means there is only one type in the union.
-        return isValueType(firstMember) && hasFillerValue(firstMember);
-    }
+    private static boolean checkFillerValue(BType type, List<Type> unAnalyzedTypes) {
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:2884]: Shall we introduce a separate method `checkFillerValue` method for `BUnionType` instead? We have several other similar methods too.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java [hunk 5]
```diff
@@ -53,21 +64,219 @@ public BUnionType() {
...
+    public boolean isNilable() {
+        if (memberTypes == null || memberTypes.isEmpty()) {
+            return true;
+        }
+
+        if (this.resolving) {
+            return false;
+        }
+
+        // TODO: use the flag
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java:245]: Can we do this change now or should we create an issue?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java [hunk 6]
```diff
@@ -53,21 +64,219 @@ public BUnionType() {
...
+        if (nullable == null) {
+            nullable = checkNillable(memberTypes);
+        }
+        return nullable;
+    }
+
+    private boolean checkNillable(List<Type> memberTypes) {
+        this.resolving = true;
+        for (Type t : memberTypes) {
+            if (t.isNilable()) {
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java:255]: Shall we avoid single character variable names?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/ReadOnlyUtils.java [hunk 7]
```diff
@@ -238,9 +272,15 @@ private static BIntersectionType setImmutableIntersectionType(Type type, Set<Typ
...
                 BUnionType origUnionType = (BUnionType) type;
+
+
                 Type resultantImmutableType;
 
                 List<Type> readOnlyMemTypes = new ArrayList<>();
+//                resultantImmutableType = new BUnionType(null, true);
+//                BUnionType unionImmutableType = (BUnionType) resultantImmutableType;
+//                unionImmutableType.setMemberTypes(readOnlyMemTypes);
+//                unionImmutableType.isCyclic = origUnionType.isCyclic;
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/ReadOnlyUtils.java:283]: Can we remove this?

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/elements/MarkdownDocAttachment.java [hunk 8]
```diff
@@ -32,8 +32,12 @@
     public List<Parameter> parameters;
     public String returnValueDescription;
 
-    public MarkdownDocAttachment() {
-        this.parameters = new ArrayList<>();
+//    public MarkdownDocAttachment() {
+//        this.parameters = new ArrayList<>();
+//    }
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/model/elements/MarkdownDocAttachment.java:37]: Shall we remove this?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/CompilerDriver.java [hunk 9]
```diff
@@ -225,7 +231,11 @@ void loadLangModules(List<PackageID> pkgIdList) {
             symbolTable.langErrorModuleSymbol = pkgLoader.loadPackageSymbol(ERROR, null, null);
         }
 
-        symResolver.reloadIntRangeType();
+        if (langLib.equals(ERROR)) {
+            symbolTable.langValueModuleSymbol = pkgLoader.loadPackageSymbol(VALUE, null, null);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/CompilerDriver.java:235]: Why do we need this? Error shouldn't depend on value, right?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsAnydataUniqueVisitor.java [hunk 10]
```diff
@@ -0,0 +1,343 @@
+package org.wso2.ballerinalang.compiler.semantics.analyzer;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsAnydataUniqueVisitor.java:1]: Missing licence header. Please check other new files too.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsAnydataUniqueVisitor.java [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangExpression;
+import org.wso2.ballerinalang.compiler.util.TypeTags;
+
+import java.util.HashSet;
+
+/**
+ * IsAnydataUniqueVisitor to check if a type is anydata.
+ *
+ * This is introduced to handle cyclic unions.
+ * @since slp4
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsAnydataUniqueVisitor.java:44]:
```suggestion
 * @since 2.0.0
```

Please check wherever applicable.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsAnydataUniqueVisitor.java [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+        visited = new HashSet<>();
+        isAnydata = true;
+    }
+
+    public IsAnydataUniqueVisitor(HashSet<BType> visited) {
+        this.visited = visited;
+        isAnydata = true;
+    }
+
+    private boolean isanydata(BType type) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsAnydataUniqueVisitor.java:61]:
```suggestion
    private boolean isAnydata(BType type) {
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/builtinoperations/CloneOperationTest.java [hunk 13]
```diff
@@ -317,11 +317,11 @@ public void testCloneMap() {
...
     public void testCloneNegative() {
-        Assert.assertEquals(negativeResult.getErrorCount(), 3);
+        Assert.assertEquals(negativeResult.getErrorCount(), 1);
         BAssertUtil.validateError(negativeResult, 0, "too many arguments in call to 'clone()'", 19, 13);
-        BAssertUtil.validateError(negativeResult, 1, "incompatible types: expected 'anydata', found 'typedesc<int>'",
-                                  24, 23);
-        BAssertUtil.validateError(negativeResult, 2, "incompatible types: expected 'anydata', found 'error'", 29, 15);
+//        BAssertUtil.validateError(negativeResult, 1, "incompatible types: expected 'anydata', found 'typedesc<int>'",
+//                                  24, 23);
+//        BAssertUtil.validateError(negativeResult, 2, "incompatible types: expected 'anydata', found 'error'", 29, 15);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/builtinoperations/CloneOperationTest.java:324]: Why are we commenting these out? These are valid errors, right?

File: stdlib/mime/src/main/ballerina/src/mime/natives.bal [hunk 14]
```diff
@@ -456,7 +456,8 @@ public class Entity {
     #
     # + return - All header names as a `string[]`
     public function getHeaderNames() returns @tainted string[] {
-        return self.headerNames.clone();
+        return <string[]>self.headerNames.clone();
```
Comment (MaryamZi) [inline stdlib/mime/src/main/ballerina/src/mime/natives.bal:459]: Unnecessary changes?

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/TypeParamTest.java [hunk 15]
```diff
@@ -40,8 +40,6 @@ public void testTypeParamNegative() {
         CompileResult result = BCompileUtil.compile("test-src/type-param/type_param_test_negative.bal");
         int err = 0;
         BAssertUtil.validateError(result, err++, "incompatible types: expected 'boolean[]', found 'int[]'", 21, 20);
-        BAssertUtil.validateError(result, err++, "incompatible types: expected 'anydata', found 'function (string) " +
```
Comment (MaryamZi) [inline langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/TypeParamTest.java:43]: This too is a valid error?

File: langlib/lang.error/src/main/ballerina/error.bal [hunk 16]
```diff
@@ -15,10 +15,14 @@
 // under the License.
 
 import ballerina/java;
+import ballerina/lang.'value as value;
+
+# The type of value to which `clone` and `cloneReadOnly` can be applied.
+type __Cloneable1 readonly|xml|__Cloneable1[]|map<__Cloneable1>|table<map<__Cloneable1>>;
```
Comment (MaryamZi) [inline langlib/lang.error/src/main/ballerina/error.bal:21]: There's a duplicate Cloneable definition in the spec for the error module. 

https://github.com/ballerina-platform/ballerina-spec/blob/master/lang/lib/error.bal#L21

Comment (KRVPerera) [reply langlib/lang.error/src/main/ballerina/error.bal:21]: Yes, but we cannot create a copy because `Cloneable` is already defined as a public type. I am getting redefinition errors.

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/TypeParamTest.java (hunk 15 above)
Comment (KRVPerera) [reply langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/TypeParamTest.java:43]: `readonly` support "function" type right. `readonly` being part of `Cloneable` shouldn't we support it now.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java (hunk 5 above)
Comment (KRVPerera) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java:245]: This is an existing TODO note. Let's create an issue.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/core/model/types/BUnionType.java (hunk 1 above)
Comment (KRVPerera) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/core/model/types/BUnionType.java:100]: Not necessarily if someone creates one programmatically.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/builtinoperations/CloneOperationTest.java (hunk 13 above)
Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/builtinoperations/CloneOperationTest.java:324]: Since `Cloneable` support `readonly` we support type we support `typedesc` and `error` as selectively immutable basic types. These are not errors now right ?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/PredefinedTypes.java (hunk 3 above)
Comment (KRVPerera) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/PredefinedTypes.java:205]: added 632ec683986ae5b2c31e5c9baf27b505629cf2c3

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/builtinoperations/CloneOperationTest.java (hunk 13 above)
Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/builtinoperations/CloneOperationTest.java:324]: Ah yes, my bad. Forgot we were changing the param type for these methods.

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/builtinoperations/CloneOperationTest.java:324]: We should remove these though. Instead of commenting them out.

File: docs/bir-spec/src/main/resources/kaitai/bir.ksy [hunk 17]
```diff
@@ -273,6 +273,14 @@ types:
         type: s4
   type_union:
     seq:
+      - id: is_cyclic
+        type: u1
+      - id: has_name
+        type: s1
+      - id: pkd_id_cp_index
```
Comment (rdhananjaya) [inline docs/bir-spec/src/main/resources/kaitai/bir.ksy:280]: I belive this one and the below one need `if` condition. `if:  has_name == 1`

File: bvm/ballerina-core/src/main/java/org/ballerinalang/core/model/types/BUnionType.java (hunk 2 above)
Comment (KRVPerera) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/core/model/types/BUnionType.java:105]: I have added `setMemberTypes` which returns if members list is null given to it. But looks like there is no way for this to happen now. But shall we keep this for now. I think if members are empty we don't need to create a member list at runtime. An example is below.
I would like to create an issue for this. What do you think?

```Ballerina
type EmptyUnion never|never;
EmptyUnion x = 3;
```

```LogTalk
ERROR [t2.bal:(2:16,2:17)] incompatible types: expected '', found 'int'
```

**Review by rdhananjaya — APPROVED** [review]

File: stdlib/mime/src/main/ballerina/src/mime/natives.bal [hunk 18]
```diff
@@ -456,7 +456,7 @@ public class Entity {
     #
     # + return - All header names as a `string[]`
     public function getHeaderNames() returns @tainted string[] {
-        return <string[]>self.headerNames.clone();
+        return self.headerNames.clone();
         //return self.headerNames.clone();
```
Comment (KRVPerera) [inline stdlib/mime/src/main/ballerina/src/mime/natives.bal:460]: need to cleanup

File: stdlib/mime/src/main/ballerina/src/mime/natives.bal (hunk 14 above)
Comment (KRVPerera) [reply stdlib/mime/src/main/ballerina/src/mime/natives.bal]: reverted

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsAnydataUniqueVisitor.java (hunk 10 above)
Comment (riyafa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsAnydataUniqueVisitor.java:1]: These files are still missing license headers
